R/expectations-matches.R

```diff
@@ -126,39 +129,58 @@ prints_text <- function(regexp, ...) {
 
 #' @export
 #' @rdname matching-expectations
-expect_error <- function(object, regexp = NULL, ..., info = NULL,
-  label = NULL) {
+expect_error <- function(object, regexp = NULL, expected_class = NULL, ..., info = NULL,
+                         label = NULL) {
   if (is.null(label)) {
     label <- find_expr("object")
   }
-  expect_that(object, throws_error(regexp, ...), info = info, label = label)
+  expect_that(object, 
+              throws_error(regexp, expected_class = expected_class, ...),
+              info = info, label = label)
 }
 #' @export
 #' @rdname oldskool
-throws_error <- function(regexp = NULL, ...) {
+throws_error <- function(regexp = NULL, expected_class = NULL, ...) {
+  if (identical(regexp, NA) && !is.null(expected_class)) {
+    stop("If regexp argument is NA (expecting no error), class should be NULL")
+  }
   function(expr) {
-    res <- try(force(expr), TRUE)
-
-    if (inherits(res, "try-error")) {
-      errors <- as.character(res)
-    } else {
-      errors <- character()
-    }
-
+    error <- tryCatch({
+      expr
+      NULL
+    },
+    error = function(e) {
+      e
+    })
+    
+    paste_with_commas <- function(x, y) paste(x, y, sep=", ")
+    expected_classes_as_string <- Reduce(paste_with_commas, expected_class)
+    no_error_string <- "no errors raised"
+    
     if (identical(regexp, NA)) {
       expectation(
-        length(errors) == 0,
-        paste0("expected no errors:\n", paste("* ", errors, collapse = "\n")),
-        "no errors raised"
+        is.null(error),
+        paste0("expected no errors:\n", as.character(error),
+               no_error_string),
+        success_msg = no_error_string
```

Comment:
`no_error_string` is a success message when there is no error (as expected), but a failure message when we expect an error but there's none.

Btw, I think `regexp=NA` meaning we expect no error is part of what makes this confusing. It looks to me like that behavior was documented months ago but only actually worked after @krlmlr's fixes.